lib/tran.c

```diff
@@ -265,36 +265,98 @@ recv_version(vfu_ctx_t *vfu_ctx, uint16_t *msg_idp,
     return 0;
 }
 
+/*
+ * A json_object_object_add wrapper that takes ownership of *val unconditionally:
+ * Resets *val to NULL and makes sure *val gets dropped, even when an error
+ * occurs. Assumes key is new and a constant.
+ */
+static int
+json_add(struct json_object *jso, const char *key, struct json_object **val)
+{
+    int ret = 0;
+
+#if JSON_C_MAJOR_VERSION > 0 || JSON_C_MINOR_VERSION >= 13
+    ret = json_object_object_add_ex(jso, key, *val,
+                                    JSON_C_OBJECT_ADD_KEY_IS_NEW |
+                                    JSON_C_OBJECT_KEY_IS_CONSTANT);
+#else
+    /* Earlier versions will abort() on allocation failure. */
+    json_object_object_add(jso, key, *val);
+#endif
+
+    if (ret < 0) {
+        json_object_put(*val);
+    }
+    *val = NULL;
+    return ret;
+}
+
+static json_object* create_caps_object(vfu_ctx_t *vfu_ctx) {
+    struct json_object *jo_migration = NULL;
+    struct json_object *jo_caps = NULL;
+    struct json_object *jo_top = NULL;
+    struct json_object *jo_tmp = NULL;
+
+    if ((jo_caps = json_object_new_object()) == NULL) {
+        goto error;
+    }
+
+    if ((jo_tmp = json_object_new_int64(SERVER_MAX_FDS)) == NULL ||
```

Comment:
consider a json_add_int64() on top of json_add() since you do this several times?

Reply:
Good idea, will do.